File: 202-happy number-快乐数/202-happy number-快乐数.java

class Solution {
    private int digitSquare(int n){
        int sum = 0;
        while(n != 0){
            int digit = n % 10;
            sum += digit * digit;
            n /= 10;
        }
        return sum;
    }
    
    public boolean isHappy(int n) {
        int slow = n;
        int fast = digitSquare(n);
        while(fast != 1 && slow != fast){
            slow = digitSquare(slow);
            fast = digitSquare(digitSquare(fast));
        }
        return fast == 1;
    }
}

File: 202-happy number-快乐数/202-happy number-快乐数.js

/**
 * @param {number} n
 * @return {boolean}
 */
 var isHappy = function(n) {
    let set = new Set()
    while(true){
        powNum = digitSquare(n)
        if(powNum === 1){
            return true
        }
        if(set.has(powNum)){
            return false
        }else{
            set.add(powNum)
        }
        n = powNum
    }
};

var digitSquare = function(n){
    let sum = 0
    while(n !== 0){
        let digit = n%10
        sum += digit**2
        n = Math.floor(n/10)
    }
    return sum
}